Improve error handling and warn before disconnecting

medape opened the pull request:
The two first commits should be self-explanatory. The last one... you can guess.

Besides making Yoke a bit more pleasant and easier to use, this also simplifies the work of `service.py`. If devices are forbidden from changing while they're connected, the PC doesn't have to check that as often, and the main loop can be kept more tight.

To that end, some collaboration from the app is necessary. The app will send three rows of 0xFF bytes, exactly as long as the last status report, before a deliberate disconnection.

Also, one of our error messages in English is a bit more polite now.

Commit:
Slightly improve error handling

Yoke app now translates to plain English a couple more errors
(webserver not found, connectivity lost, connection refused before starting,
connection refused when Yoke was closed on PC).

Commit:
Yoke now warns if the user deliberately closes the connection

This simplifies some checks in service.py. It should also allow for
users to come back to the app to recover the connection if they
accidentally exited the app, but this requires more tests and raising
the timeout limit, so nothing has been done in this direction yet.

An important change was made to accomodate this: THE YOKE APP NOW
ACTUALLY ENFORCES A CONSISTENT POLL RATE. (Previously it enforced only
a minimum poll rate.) This can be lowered or made customizable in the
future.

Commit:
nulls for EVERYBODY!

pzmarzly commented on app/src/main/java/com/simonramstedt/yoke/YokeActivity.java:
```diff
@@ -709,23 +712,43 @@ public void onStopDiscoveryFailed(String serviceType, int errorCode) {
+    // closeConnection(false) closes the connection with no warning: it's intended for both
+    // accidental disconecttions (like exiting the app) and error states (where it doesn't make
+    // sense to warn the client and doing so would probably send to a null socket or cause more errors.)
+    private void closeConnection(boolean warn) {
```
I'm not sure whether it's better to keep this function this way, or to split it into 2 functions (`requestDisconnect` and `closeSocket`). Boolean flags are usually not that clear:

```java
this.requestDisconnect();
this.closeSocket();
```

vs

```java
this.closeConnection(true);
```

Plus, the comment makes it sound like these are 2 separate functions, though that's just my opinion.

medape replied:
Your proposal is much more clear (and cleaner), I'll push a commit to adopt it.

I'll also push a commit to both PRs to move the disconnection pattern to the JavaScript side. It's much more useful and tunable than hardcoding it in the app.

medape replied:
Commits 3c9459460ea07e9c21b9fc15112814ab28319e6a and 6b4cc04693b53747d4160156d593e1d3a0335d15

And commit 199e5d5b9469b91b523aa715e49e8f2b0c0ce9cc on the other repo.

Commit:
Clarify the code

Split closeConnection(boolean warn) into requestDisconnect()
and closeSocket()

Commit:
Move the disconnection pattern to JavaScript

Some behaviour is still hard-coded on the app (the pattern will
be sent three times) but the actual pattern is easily customizable
from the JavaScript side.
